File: mac-tracker/Sources/WorkTrackerTracker/ActivityQueue.swift
import Foundation

/// A small local queue of not-yet-sent activity timestamps, persisted to
/// disk so a network blip (or the tracker quitting) doesn't lose events —
/// see API_CONTRACT.md's note on why batched event posting exists. Only
/// entries the server has actually accepted are ever removed.
///
/// Behavior during a sustained server outage (kept in sync with the Windows
/// tracker's `ActivityQueue.cs`):
///  - `flush()` sends in chunks of at most `chunkSize` timestamps, well
///    under the server's per-request cap, and removes each chunk as soon
///    as it's acknowledged — so a huge backlog drains in pieces instead of
///    being rejected as one oversized request.
///  - Only the entries that were actually sent are removed; timestamps
///    enqueued while a request was in flight stay queued.
///  - Repeated failures back off exponentially instead of letting the
///    caller's fixed-interval timer hammer the endpoint.
///  - A chunk the server rejects as invalid (HTTP 400/413/422) is dropped
///    rather than retried forever, so it can't block everything behind it.
///  - `pending` is capped very generously; once full the oldest entries
///    are dropped and counted in `droppedEventCount`.
///
/// Thread-safety: `enqueue` runs on the main thread, `flush` on whatever
/// executor its `Task` lands on; shared state is guarded by `lock`, which
/// is never held across an `await`.
final class ActivityQueue {
    /// Timestamps are ~30 bytes each on disk, so 100k entries is ~3 MB —
    /// enough for weeks of activity at the default poll interval — while
    /// still bounding memory/disk use if the server is gone for good.
    static let defaultMaxPendingCount = 100_000

    /// Timestamps per request. Must stay at or below the server's
    /// per-request cap (`MAX_EVENTS_PER_REQUEST` in src/server/app.ts).
    static let defaultChunkSize = 1000

    /// Base delay before retrying after the *first* consecutive flush
    /// failure; doubles with each further consecutive failure (see
    /// `backoffInterval(forConsecutiveFailures:)`), capped at
    /// `defaultMaxBackoffIntervalSeconds`.
    static let defaultBaseBackoffIntervalSeconds: TimeInterval = 15

    /// Upper bound on the backoff delay, so a server outage lasting hours
    /// still gets retried a few times an hour rather than essentially never.
    static let defaultMaxBackoffIntervalSeconds: TimeInterval = 300

    /// Pending changes are batched to disk at most this often, to avoid
    /// rewriting the whole queue file on every single `enqueue()` call.
    /// Losing the last few seconds of unpersisted state in a crash (as
    /// opposed to a graceful quit, which calls `persistNow()`) is an
    /// acceptable trade-off for activity timestamps.
    static let defaultPersistDebounceIntervalSeconds: TimeInterval = 5

    private var pending: [Date]
    private let storageURL: URL
    private let dateFormatter: ISO8601DateFormatter
    private let maxPendingCount: Int
    private let chunkSize: Int
    private let baseBackoffInterval: TimeInterval
    private let maxBackoffInterval: TimeInterval
    private let persistDebounceInterval: TimeInterval
    private let now: () -> Date

    private var lastPersistedAt: Date?
    private var hasUnpersistedChanges = false
    private var consecutiveFailureCount = 0
    private var nextAllowedFlushAt: Date?
    private var isFlushing = false
    private let lock = NSLock()

    /// Monotonic sequence number of `pending[0]`. Lets a flush remove
    /// exactly the entries it sent even if cap eviction shifted the front
    /// of the queue while the request was in flight.
    private var headSequence = 0

    /// Human-readable reason the most recent flush failed, or nil after a
    /// success. Shown in the status menu.
    private(set) var lastError: String?

    /// Total entries discarded this session: evicted by the cap, or
    /// rejected by the server as invalid.
    private(set) var droppedEventCount = 0

    /// When the last batch was actually accepted by the server — not just
    /// attempted. Persisted across restarts (in the same queue file) so the
    /// status item shows an accurate value immediately after launch, not
    /// "Never synced" just because this process hasn't flushed yet.
    private(set) var lastSuccessfulSyncAt: Date?

    init(
        storageURL: URL,
        maxPendingCount: Int = ActivityQueue.defaultMaxPendingCount,
        chunkSize: Int = ActivityQueue.defaultChunkSize,
        baseBackoffInterval: TimeInterval = ActivityQueue.defaultBaseBackoffIntervalSeconds,
        maxBackoffInterval: TimeInterval = ActivityQueue.defaultMaxBackoffIntervalSeconds,
        persistDebounceInterval: TimeInterval = ActivityQueue.defaultPersistDebounceIntervalSeconds,
        now: @escaping () -> Date = Date.init
    ) {
        self.storageURL = storageURL
        self.maxPendingCount = maxPendingCount
        self.chunkSize = max(1, chunkSize)
        self.baseBackoffInterval = baseBackoffInterval
        self.maxBackoffInterval = maxBackoffInterval
        self.persistDebounceInterval = persistDebounceInterval
        self.now = now
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        self.dateFormatter = formatter
        let loaded = Self.load(from: storageURL, using: formatter)
        self.pending = loaded.pending
        self.lastSuccessfulSyncAt = loaded.lastSuccessfulSyncAt
    }

    var pendingCount: Int { lock.withLock { pending.count } }

    /// Exposed for tests to inspect which timestamps survived cap eviction;
    /// production callers only need `pendingCount`.
    var pendingTimestamps: [Date] { lock.withLock { pending } }

    /// Number of flush attempts that have failed in a row since the last
    /// success (or since the queue was created). Exposed for tests.
    var currentConsecutiveFailureCount: Int { lock.withLock { consecutiveFailureCount } }

    func enqueue(_ date: Date) {
        lock.withLock {
            pending.append(date)
            if pending.count > maxPendingCount {
                let overflow = pending.count - maxPendingCount
                pending.removeFirst(overflow)
                headSequence += overflow
                droppedEventCount += overflow
            }
            hasUnpersistedChanges = true
            persistIfDebounceElapsed()
        }
    }

    /// Sends everything pending, one chunk at a time, stopping at the first
    /// failure. A call that arrives while another flush is running, or
    /// before `nextAllowedFlushAt`, is a cheap no-op, so the caller's
    /// fixed-interval timer can call this on every tick unconditionally.
    func flush(client: EventsAPIClient, serverBaseURL: String, apiKey: String) async {
        guard beginFlush() else { return }
        defer { lock.withLock { isFlushing = false } }

        while let chunk = nextChunk() {
            do {
                try await client.postEvents(chunk.timestamps, serverBaseURL: serverBaseURL, apiKey: apiKey)
                completeChunk(endSequence: chunk.endSequence, rejected: false)
            } catch let error as APIClientError where error.isPermanentRejection {
                // The server understood the request and refuses this data;
                // retrying the same chunk would block the queue forever.
                completeChunk(endSequence: chunk.endSequence, rejected: true)
                lock.withLock { lastError = "Server rejected \(chunk.timestamps.count) event(s) as invalid" }
            } catch {
                recordFailure(error)
                return
            }
        }
    }

    private func beginFlush() -> Bool {
        lock.withLock {
            guard !isFlushing, !pending.isEmpty else { return false }
            if let nextAllowedFlushAt, now() < nextAllowedFlushAt { return false }
            isFlushing = true
            return true
        }
    }

    private func nextChunk() -> (timestamps: [Date], endSequence: Int)? {
        lock.withLock {
            guard !pending.isEmpty else { return nil }
            let chunk = Array(pending.prefix(chunkSize))
            return (chunk, headSequence + chunk.count)
        }
    }

    /// Removes exactly the entries up to `endSequence` — never more, so
    /// timestamps enqueued (or already evicted) meanwhile are untouched.
    private func completeChunk(endSequence: Int, rejected: Bool) {
        lock.withLock {
            let removable = min(max(0, endSequence - headSequence), pending.count)
            pending.removeFirst(removable)
            headSequence += removable
            if rejected {
                droppedEventCount += removable
            } else {
                consecutiveFailureCount = 0
                nextAllowedFlushAt = nil
                lastError = nil
                lastSuccessfulSyncAt = now()
            }
            hasUnpersistedChanges = true
            persist()
        }
    }

    private func recordFailure(_ error: Error) {
        lock.withLock {
            consecutiveFailureCount += 1
            let delay = Self.backoffInterval(
                forConsecutiveFailures: consecutiveFailureCount,
                base: baseBackoffInterval,
                max: maxBackoffInterval
            )
            nextAllowedFlushAt = now().addingTimeInterval(delay)
            lastError = Self.describe(error)
        }
    }

    private static func describe(_ error: Error) -> String {
        switch error as? APIClientError {
        case .unauthorized: return "API key invalid or revoked"
        case .invalidServerURL: return "Invalid server URL"
        case .requestFailed(let status) where status > 0: return "Server error (HTTP \(status))"
        default: return "Server unreachable"
        }
    }

    /// Doubles the delay with each consecutive failure (1st failure: base,
    /// 2nd: 2x base, 3rd: 4x base, ...), capped at `max`.
    static func backoffInterval(forConsecutiveFailures count: Int, base: TimeInterval, max: TimeInterval) -> TimeInterval {
        guard count > 0 else { return 0 }
        let multiplier = pow(2.0, Double(count - 1))
        return Swift.min(base * multiplier, max)
    }

    /// Forces any pending in-memory changes to disk immediately, bypassing
    /// the debounce window. Call this on graceful app termination so the
    /// worst case data loss is limited to a crash between debounce windows,
    /// not a normal quit.
    func persistNow() {
        lock.withLock {
            guard hasUnpersistedChanges else { return }
            persist()
        }
    }

    private func persistIfDebounceElapsed() {
        if let lastPersistedAt, now().timeIntervalSince(lastPersistedAt) < persistDebounceInterval {
            return
        }
        persist()
    }

    /// Caller must hold `lock`. Writes atomically (temp file + rename), so
    /// a crash mid-write can't leave a truncated queue file. A failed write
    /// leaves `hasUnpersistedChanges` set so the next attempt retries.
    private func persist() {
        let state = PersistedState(
            pending: pending.map { dateFormatter.string(from: $0) },
            lastSuccessfulSyncAt: lastSuccessfulSyncAt.map { dateFormatter.string(from: $0) }
        )
        do {
            let data = try JSONEncoder().encode(state)
            try FileManager.default.createDirectory(
                at: storageURL.deletingLastPathComponent(), withIntermediateDirectories: true
            )
            try data.write(to: storageURL, options: .atomic)
            hasUnpersistedChanges = false
        } catch {
            hasUnpersistedChanges = true
        }
        lastPersistedAt = now()
    }

    /// The on-disk shape. Kept separate from `[Date]`/`Date` so it can be
    /// `Codable` without teaching `Date` to round-trip through the same
    /// `ISO8601DateFormatter` this class already uses elsewhere.
    private struct PersistedState: Codable {
        var pending: [String]
        var lastSuccessfulSyncAt: String?
    }

    private static func load(from url: URL, using formatter: ISO8601DateFormatter) -> (pending: [Date], lastSuccessfulSyncAt: Date?) {
        guard let data = try? Data(contentsOf: url) else { return ([], nil) }

        if let state = try? JSONDecoder().decode(PersistedState.self, from: data) {
            let pending = state.pending.compactMap { formatter.date(from: $0) }
            let lastSync = state.lastSuccessfulSyncAt.flatMap { formatter.date(from: $0) }
            return (pending, lastSync)
        }

        // Older queue files are a bare `[String]` of pending timestamps
        // (no lastSuccessfulSyncAt yet) — fall back to that shape so
        // upgrading doesn't drop an existing queue.
        if let strings = try? JSONDecoder().decode([String].self, from: data) {
            return (strings.compactMap { formatter.date(from: $0) }, nil)
        }

        // Neither shape parsed: keep the unreadable file for manual
        // recovery instead of silently overwriting it with an empty queue.
        let quarantine = url.appendingPathExtension("corrupt")
        try? FileManager.default.removeItem(at: quarantine)
        try? FileManager.default.moveItem(at: url, to: quarantine)
        return ([], nil)
    }
}

File: windows-tracker/src/WorkTrackerTracker.Core/ActivityQueue.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace WorkTrackerTracker.Core;

/// <summary>
/// A small local queue of not-yet-sent activity timestamps, persisted to
/// disk so a network blip (or the tracker quitting) doesn't lose events —
/// see docs/API_CONTRACT.md's note on why batched event posting exists.
/// Only entries the server has actually accepted are ever removed.
/// </summary>
/// <remarks>
/// Behavior during a sustained server outage (kept in sync with the Mac
/// tracker's ActivityQueue.swift):
/// <list type="bullet">
/// <item>FlushAsync sends in chunks of at most <c>chunkSize</c> timestamps,
/// under the server's per-request cap, and removes each chunk as soon as
/// it's acknowledged, so a huge backlog drains in pieces.</item>
/// <item>Only the entries that were actually sent are removed; timestamps
/// enqueued while a request was in flight stay queued.</item>
/// <item>Repeated failures back off exponentially.</item>
/// <item>A chunk the server rejects as invalid (HTTP 400/413/422) is dropped
/// rather than retried forever, so it can't block everything behind it.</item>
/// <item>The queue is capped very generously; once full the oldest entries
/// are dropped and counted in <see cref="DroppedEventCount"/>.</item>
/// </list>
/// Thread-safety: Enqueue runs on the UI thread, while the continuation of
/// FlushAsync may resume on a pool thread (ConfigureAwait(false)); shared
/// state is guarded by a lock that is never held across an await.
/// </remarks>
public sealed class ActivityQueue
{
    /// <summary>~30 bytes per entry on disk: 100k entries is ~3 MB.</summary>
    public const int DefaultMaxPendingCount = 100_000;

    /// <summary>Must stay at or below the server's MAX_EVENTS_PER_REQUEST (src/server/app.ts).</summary>
    public const int DefaultChunkSize = 1000;

    public static readonly TimeSpan DefaultBaseBackoff = TimeSpan.FromSeconds(15);
    public static readonly TimeSpan DefaultMaxBackoff = TimeSpan.FromSeconds(300);

    /// <summary>
    /// Pending changes are written to disk at most this often; a graceful
    /// exit calls <see cref="PersistNow"/>.
    /// </summary>
    public static readonly TimeSpan DefaultPersistDebounce = TimeSpan.FromSeconds(5);

    private readonly object _lock = new();
    private readonly string _storagePath;
    private readonly Func<DateTimeOffset> _now;
    private readonly int _maxPendingCount;
    private readonly int _chunkSize;
    private readonly TimeSpan _baseBackoff;
    private readonly TimeSpan _maxBackoff;
    private readonly TimeSpan _persistDebounce;

    private List<DateTimeOffset> _pending;
    private long _headSequence; // sequence number of _pending[0]
    private DateTimeOffset? _lastPersistedAt;
    private bool _hasUnpersistedChanges;
    private int _consecutiveFailureCount;
    private DateTimeOffset? _nextAllowedFlushAt;
    private bool _isFlushing;
    private DateTimeOffset? _lastSuccessfulSyncAt;
    private string? _lastError;
    private int _droppedEventCount;

    public ActivityQueue(
        string storagePath,
        Func<DateTimeOffset>? now = null,
        int maxPendingCount = DefaultMaxPendingCount,
        int chunkSize = DefaultChunkSize,
        TimeSpan? baseBackoff = null,
        TimeSpan? maxBackoff = null,
        TimeSpan? persistDebounce = null)
    {
        _storagePath = storagePath;
        _now = now ?? (() => DateTimeOffset.UtcNow);
        _maxPendingCount = maxPendingCount;
        _chunkSize = Math.Max(1, chunkSize);
        _baseBackoff = baseBackoff ?? DefaultBaseBackoff;
        _maxBackoff = maxBackoff ?? DefaultMaxBackoff;
        _persistDebounce = persistDebounce ?? DefaultPersistDebounce;

        var loaded = Load(storagePath);
        _pending = loaded.Pending;
        _lastSuccessfulSyncAt = loaded.LastSuccessfulSyncAt;
    }

    public int PendingCount { get { lock (_lock) return _pending.Count; } }

    /// <summary>Snapshot of the queued timestamps; exposed for tests.</summary>
    public IReadOnlyList<DateTimeOffset> PendingTimestamps { get { lock (_lock) return _pending.ToArray(); } }

    /// <summary>
    /// When the last batch was actually accepted by the server — not just
    /// attempted. Persisted across restarts (in the same queue file).
    /// </summary>
    public DateTimeOffset? LastSuccessfulSyncAt { get { lock (_lock) return _lastSuccessfulSyncAt; } }

    /// <summary>Why the most recent flush failed, or null after a success.</summary>
    public string? LastError { get { lock (_lock) return _lastError; } }

    /// <summary>Entries discarded this session: evicted by the cap or rejected as invalid.</summary>
    public int DroppedEventCount { get { lock (_lock) return _droppedEventCount; } }

    /// <summary>Consecutive failed flushes since the last success; exposed for tests.</summary>
    public int ConsecutiveFailureCount { get { lock (_lock) return _consecutiveFailureCount; } }

    public void Enqueue(DateTimeOffset timestamp)
    {
        lock (_lock)
        {
            _pending.Add(timestamp);
            if (_pending.Count > _maxPendingCount)
            {
                var overflow = _pending.Count - _maxPendingCount;
                _pending.RemoveRange(0, overflow);
                _headSequence += overflow;
                _droppedEventCount += overflow;
            }
            _hasUnpersistedChanges = true;
            PersistIfDebounceElapsed();
        }
    }

    /// <summary>
    /// Sends everything pending, one chunk at a time, stopping at the first
    /// failure. A call that arrives while another flush is running, or
    /// before the backoff window has elapsed, is a cheap no-op, so the
    /// caller's fixed-interval timer can call this on every tick.
    /// </summary>
    public async Task FlushAsync(IEventsApiClient client, string serverBaseUrl, string apiKey, CancellationToken cancellationToken = default)
    {
        if (!BeginFlush())
        {
            return;
        }

        try
        {
            while (NextChunk() is { } chunk)
            {
                try
                {
                    await client.PostEventsAsync(chunk.Timestamps, serverBaseUrl, apiKey, cancellationToken).ConfigureAwait(false);
                    CompleteChunk(chunk.EndSequence, rejected: false);
                }
                catch (ApiClientException ex) when (ex.IsPermanentRejection)
                {
                    // The server understood the request and refuses this
                    // data; retrying the same chunk would block the queue.
                    CompleteChunk(chunk.EndSequence, rejected: true);
                    lock (_lock) _lastError = $"Server rejected {chunk.Timestamps.Length} event(s) as invalid";
                }
                catch (Exception ex) when (ex is not OperationCanceledException || !cancellationToken.IsCancellationRequested)
                {
                    RecordFailure(ex);
                    return;
                }
            }
        }
        finally
        {
            lock (_lock) _isFlushing = false;
        }
    }

    private bool BeginFlush()
    {
        lock (_lock)
        {
            if (_isFlushing || _pending.Count == 0) return false;
            if (_nextAllowedFlushAt is { } next && _now() < next) return false;
            _isFlushing = true;
            return true;
        }
    }

    private (DateTimeOffset[] Timestamps, long EndSequence)? NextChunk()
    {
        lock (_lock)
        {
            if (_pending.Count == 0) return null;
            var chunk = _pending.Take(_chunkSize).ToArray();
            return (chunk, _headSequence + chunk.Length);
        }
    }

    /// <summary>Removes exactly the entries up to <paramref name="endSequence"/>, never more.</summary>
    private void CompleteChunk(long endSequence, bool rejected)
    {
        lock (_lock)
        {
            var removable = (int)Math.Min(Math.Max(0, endSequence - _headSequence), _pending.Count);
            _pending.RemoveRange(0, removable);
            _headSequence += removable;
            if (rejected)
            {
                _droppedEventCount += removable;
            }
            else
            {
                _consecutiveFailureCount = 0;
                _nextAllowedFlushAt = null;
                _lastError = null;
                _lastSuccessfulSyncAt = _now();
            }
            _hasUnpersistedChanges = true;
            Persist();
        }
    }

    private void RecordFailure(Exception error)
    {
        lock (_lock)
        {
            _consecutiveFailureCount++;
            _nextAllowedFlushAt = _now() + BackoffInterval(_consecutiveFailureCount, _baseBackoff, _maxBackoff);
            _lastError = Describe(error);
        }
    }

    private static string Describe(Exception error) => error switch
    {
        ApiClientException { IsUnauthorized: true } => "API key invalid or revoked",
        ApiClientException { StatusCode: { } status } => $"Server error (HTTP {status})",
        ApiClientException ex => ex.Message,
        _ => "Server unreachable",
    };

    /// <summary>Doubles with each consecutive failure (base, 2x, 4x, ...), capped at <paramref name="max"/>.</summary>
    public static TimeSpan BackoffInterval(int consecutiveFailures, TimeSpan baseInterval, TimeSpan max)
    {
        if (consecutiveFailures <= 0) return TimeSpan.Zero;
        var seconds = baseInterval.TotalSeconds * Math.Pow(2, consecutiveFailures - 1);
        return TimeSpan.FromSeconds(Math.Min(seconds, max.TotalSeconds));
    }

    /// <summary>Forces pending changes to disk now, bypassing the debounce. Call on graceful exit.</summary>
    public void PersistNow()
    {
        lock (_lock)
        {
            if (_hasUnpersistedChanges) Persist();
        }
    }

    // Caller holds _lock.
    private void PersistIfDebounceElapsed()
    {
        if (_lastPersistedAt is { } last && _now() - last < _persistDebounce)
        {
            return;
        }
        Persist();
    }

    /// <summary>
    /// Caller holds _lock. Writes to a temp file and swaps it in, so a crash
    /// mid-write can't leave a truncated queue. A failed write is swallowed
    /// (the queue stays in memory and the next attempt retries) rather than
    /// crashing the tray app from inside a timer tick.
    /// </summary>
    private void Persist()
    {
        _lastPersistedAt = _now();
        try
        {
            var directory = Path.GetDirectoryName(_storagePath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            var state = new PersistedState
            {
                Pending = _pending.Select(FormatIso8601).ToArray(),
                LastSuccessfulSyncAt = _lastSuccessfulSyncAt is { } syncAt ? FormatIso8601(syncAt) : null,
            };
            var tempPath = _storagePath + ".tmp";
            File.WriteAllText(tempPath, JsonSerializer.Serialize(state));
            File.Move(tempPath, _storagePath, overwrite: true);
            _hasUnpersistedChanges = false;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            _hasUnpersistedChanges = true;
        }
    }

    private static string FormatIso8601(DateTimeOffset timestamp) =>
        timestamp.UtcDateTime.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture);

    private static (List<DateTimeOffset> Pending, DateTimeOffset? LastSuccessfulSyncAt) Load(string path)
    {
        if (!File.Exists(path))
        {
            return (new List<DateTimeOffset>(), null);
        }

        string json;
        try
        {
            json = File.ReadAllText(path);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            // Unreadable right now (e.g. locked): start empty but leave the
            // file alone — the next successful persist replaces it, and we
            // must not treat this as corruption.
            return (new List<DateTimeOffset>(), null);
        }

        try
        {
            var state = JsonSerializer.Deserialize<PersistedState>(json);
            if (state is not null)
            {
                var pending = ParseTimestamps(state.Pending);
                var lastSync = state.LastSuccessfulSyncAt is { } s && DateTimeOffset.TryParse(s, null, DateTimeStyles.RoundtripKind, out var dt)
                    ? dt
                    : (DateTimeOffset?)null;
                return (pending, lastSync);
            }
        }
        catch (JsonException)
        {
            // Fall through to the legacy shape below.
        }

        try
        {
            // Older queue files are a bare `[String]` of pending timestamps
            // (no LastSuccessfulSyncAt yet) — fall back to that shape so
            // upgrading doesn't drop an existing queue.
            var iso = JsonSerializer.Deserialize<string[]>(json) ?? Array.Empty<string>();
            return (ParseTimestamps(iso), null);
        }
        catch (JsonException)
        {
            Quarantine(path);
            return (new List<DateTimeOffset>(), null);
        }
    }

    /// <summary>Keeps an unparseable queue file for manual recovery instead of overwriting it with an empty queue.</summary>
    private static void Quarantine(string path)
    {
        try
        {
            File.Move(path, path + ".corrupt", overwrite: true);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            // Best effort.
        }
    }

    private static List<DateTimeOffset> ParseTimestamps(IEnumerable<string> iso) =>
        iso
            .Select(s => DateTimeOffset.TryParse(s, null, DateTimeStyles.RoundtripKind, out var dt) ? dt : (DateTimeOffset?)null)
            .Where(dt => dt.HasValue)
            .Select(dt => dt!.Value)
            .ToList();

    private sealed class PersistedState
    {
        [JsonPropertyName("pending")]
        public string[] Pending { get; set; } = Array.Empty<string>();

        [JsonPropertyName("lastSuccessfulSyncAt")]
        public string? LastSuccessfulSyncAt { get; set; }
    }
}
